### ISSUE TITLE
Add support for GTIN-14 encoding

### ISSUE THREAD

**Issue by kokoshneta** [biblys/isbn#25]
I use this excellent class for virtually all things ISBN-related (working for a publisher, that’s quite a lot), but there’s one area where it falls short: when shipping titles, many recipients require packing labels to include a barcode of the ISBN encoded as GTIN-14 (GS1-128).

The conversion itself is simple enough: any valid EAN can be represented as GTIN-14 by adding a single-digit prefix (most commonly 1, but any digit is valid) and recalculating the checksum. Checksum calculation for GTIN-14 also follows the same principles as for EAN (alternately multiplying digits by 3 and 1 right-to-left and getting the modulo of the sum).

In other words it should in theory be perfectly easy and straightforward to add GTIN capabilities to the class. In fact, since the underlying identifier doesn’t change, only the prefix and the checksum, it should be possible – as far as I can tell – to do this by doing nothing more than modifying the `format` and `calculateChecksum` functions and adding a property `_prefix`.

Would this be useful to anyone but me, or would it be felt to be outside the scope of the class? If it’s not felt to be out of scope, I’d make a pull request for merging.

**Comment by iwazaru** [comment]
Hi kokoshneta and thank you, I'm so glad this library is useful to you!

You're very welcome if you want to throw a pull request, I don't know the first thing about GTIN-14, but I'm sure it could be useful for others it he future! 

You're right about the need to modify the `format` and `calculateChecksum` methods, but this won't be enough. The `constructor` method where the input code parsing happens expects a 10 or 13 long string, so it should be adapted to handle a 14 long string and extract the prefix. Be sure to add some tests for your use cases. And feel free to create a draft pull request and ping me if you need any help!

**Comment by kokoshneta** [comment]
Thanks!

I don’t think it’s necessary to modify the constructor, because you’re still representing an ISBN, just encoding it in a different manner – so the underlying ISBN would remain the same. The prefix in a GTIN actually represents a ‘package-level identifier’, which I think is meant to identify different types of shipping configurations – but as far as I can tell, most people just use 1 as a default, unless they actually need to distinguish different types (e.g., single cartons, pallets, bags, etc.).

But this also means that the prefix is not an inherent part of the identifier (the ISBN) itself – it can change, depending on how you’re shipping the item. As such, it makes sense to me to think of the GTIN as being for output only. You’d create an ISBN object as usual (using your 10- or 13-digit number), and then you’d do `$isbn->format('GTIN-14', $prefix)` to get the GTIN-14 output, with the appropriate prefix included in the function call to the formatting function.

I’ll have a look at making a pull request and see if I can’t integrate it neatly. :-)

**Comment by iwazaru** [comment]
You're totally right, forget what I said about input and parsing. I thought you would also need to handle GTIN as inputs, but off course don't bother implement it if you don't need it ;)

**Comment by kokoshneta** [comment]
All right, pull request #28 – only done a little quick, manual testing, but seems to work, as far as I can tell. Not sure how exactly to figure out the PHPUnit testing (I’m sort of new to all this), but I’ve added one just in case.

**Comment by iwazaru** [comment]
[2.2.0](https://github.com/biblys/isbn/releases/tag/2.2.0) was just released with GTIN-14 formatting. Thanks for your contribution 🙏🏻

**Comment by kokoshneta** [comment]
Excellent! Glad I could help (even if I did need a little guiding). :-)
